processing/src/main/java/io/druid/query/groupby/epinephelinae/BufferArrayGrouper.java

```diff
@@ -0,0 +1,281 @@
+           (int) Math.ceil((double) cardinalityWithMissingValue / 8) +  // total used flags size
+           cardinalityWithMissingValue * recordSize;                    // total values size
+  }
+
+  public BufferArrayGrouper(
+      final Supplier<ByteBuffer> bufferSupplier,
+      final KeySerde<KeyType> keySerde,
+      final ColumnSelectorFactory columnSelectorFactory,
+      final AggregatorFactory[] aggregatorFactories,
+      final int cardinality
+  )
+  {
+    Preconditions.checkNotNull(aggregatorFactories, "aggregatorFactories");
+    Preconditions.checkArgument(cardinality > 0, "Cardinality must a non-zero positive number");
+
+    this.bufferSupplier = Preconditions.checkNotNull(bufferSupplier, "bufferSupplier");
+    this.keySerde = Preconditions.checkNotNull(keySerde, "keySerde");
+    this.aggregators = new BufferAggregator[aggregatorFactories.length];
+    this.aggregatorOffsets = new int[aggregatorFactories.length];
+    this.cardinalityWithMissingValue = cardinality + 1;
+
+    int offset = 0;
+    for (int i = 0; i < aggregatorFactories.length; i++) {
+      aggregators[i] = aggregatorFactories[i].factorizeBuffered(columnSelectorFactory);
+      aggregatorOffsets[i] = offset;
+      offset += aggregatorFactories[i].getMaxIntermediateSize();
+    }
+    recordSize = offset;
+  }
+
+  @Override
+  public void init()
+  {
+    if (!initialized) {
+      final ByteBuffer buffer = bufferSupplier.get().duplicate();
+
+      buffer.position(0);
+      buffer.limit(keySerde.keySize());
+      keyBuffer = buffer.slice();
+
+      final int usedBufferEnd = keySerde.keySize() + (int) Math.ceil((double) cardinalityWithMissingValue / 8);
+      buffer.position(keySerde.keySize());
+      buffer.limit(usedBufferEnd);
+      usedFlagBuffer = buffer.slice();
+
+      buffer.position(usedBufferEnd);
+      buffer.limit(buffer.capacity());
+      valBuffer = buffer.slice();
+
+      reset();
+
+      initialized = true;
+    }
+  }
+
+  @Override
+  public boolean isInitialized()
+  {
+    return initialized;
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key)
+  {
+    // BufferArrayGrouper is used only for dictionary-indexed single-value string dimensions.
+    // Here, the key contains the dictionary-encoded value of the grouping key
+    // and we use it as the index for the aggregation array.
+
+    final ByteBuffer fromKey = checkAndGetKeyBuffer(keySerde, key);
+    if (fromKey == null) {
+      // This may just trigger a spill and get ignored, which is ok. If it bubbles up to the user, the message will
+      // be correct.
+      return Groupers.DICTIONARY_FULL;
+    }
+
+    // The first index is reserved for missing value which is represented by -1.
+    final int dimIndex = fromKey.getInt() + 1;
+    fromKey.rewind();
+    return aggregate(key, dimIndex);
+  }
+
+  @Override
+  public AggregateResult aggregate(KeyType key, int dimIndex)
```

Comment:
The Grouper interface specifies that the second parameter here should be:

> result of {@link Groupers#hash(Object)} on the key

But this is not what's being passed in by the `aggregate(key)` method. So that means it's violating the Grouper contract, and as a result it will not work properly from within a SpillingGrouper or ConcurrentGrouper. Maybe deal with that in one of the following ways?

- Change the Grouper contract such that a grouper can provide its own key-hashing function. Some will provide `Groupers.hash` but this class will provide its own.
- Just keep the code the way it is, and document that this class doesn't quite follow the Grouper contract, so it should not be wrapped in Concurrent or SpillingGrouper.